ehr/src/org/labkey/ehr/utils/TriggerScriptHelper.java

```diff
@@ -138,6 +138,10 @@ public class TriggerScriptHelper
 
     private static final Logger _log = LogHelper.getLogger(TriggerScriptHelper.class, "Server-side validation of EHR data insert/update/deletes");
 
+    private static boolean _checkThreshold = false;
+
+    private static double _bloodLimitThreshold = 4.0; //in mL
```

Comment:
It seems like both of these should be set through some mechanism that is easily modified by the modules. The most natural thing would be to add properties to the existing addScriptOption mechanism, which is widely used, like:
```
    helper.setScriptOptions({
        allowAnyId: true,
        requiresStatusRecalc: true,
        allowDeadIds: true,
        skipIdFormatCheck: true,
        skipHousingCheck: true,
        announceAllModifiedParticipants: true
    });
```

This does mean it is set on the JS layer, and would need ot be propagated into the java helper. I would suggest defaulting to false, meaning we preserve the status quo unless a center-specific module opts in.

Reply:
you're right I was under the impression that we can't override the init, let me take a look at adding a script option instead.

Reply:
@bbimber any idea how to propagate these script options through to the java helper?

Reply:
Actually I think setCenterCustomProps might do the trick, let me see

Reply:
@bbimber I found something called setCenterCustomProps which might do the trick, it doesn't look like setScriptOptions gets propagated over

Reply:
If I understand you question I would think about it like this. There's a tradeoff:

- The way you implemented this now is to expose a setter on the java trigger helper class for the 'reportNearOverage' property and a setter to set the threshold value. For a center to opt-in to this behavior, they need to do that you did and override the core blood.js onUpsert() code with their own logic. 

- Option 2 would be to be more like how some existing support properties work, like 'allowDeadIds', 'doStandardProtocolCountValidation', etc. It would not be crazy to support properties related to the blood-near-overage behavior. If we did that, the core blood.js code would read that property, call the TriggerScriptHelper setters as appropriate, and parse the resulting messages into error/info as appropriate. Then the WNPRC module (or anyone else wanting this), would just need to register code vaguely like this:
```
EHR.Server.TriggerManager.registerHandlerForQuery(EHR.Server.TriggerManager.Events.INIT, 'study', 'blood', function(event, helper, EHR) {
        helper.setScriptOptions({
            reportBloodDrawNearOverage: true,
            bloodDrawNearOverageThreshold: 4
        });
    });
```

And then core EHR code would deal with the rest. That's probably how I would do it. The goal would be to keep as much of the codepath as reasonable shared between centers. 

That said, as long as this is well documented and explained in EHR's code, I think the split you have between EHR and WNPRC is reasonable as-is.

Reply:
it looks like setScriptOptions does not get propagated to the JS layer, but there is a setCenterCustomProps that does, so I went with trying that out.

it seems like when you do something like:

```
EHR.Server.TriggerManager.registerHandlerForQuery(EHR.Server.TriggerManager.Events.BEFORE_UPSERT, 'study', 'blood', function (helper, scriptErrors, row, oldRow) {
            helper.setCenterCustomProps({
                doWarnForBloodNearOverages: true,
                bloodNearOverageThreshold: 4.0 
            })
        });
```

introducing the JS layer for this seems to work ok, but you then encounter weird casting issues.
the when I send 4.0 as the val of bloodNearOverageThreshold, the JS interpreter thinks bloodNearOverageThreshold is an integer, so it needs to be converted back to a double for the calculation. but for example if I send a 4.5, it will get interpreted as a double.

Reply:
I dont mean to introduce a bunch of extra hurdles here, and I could see multiple ways of approaching this being fine. 

First: the reason you have having problems with setCenterCustomProps() is that this is getting passed from the JS layer to the java layer as a Map<String, String>. You could probably do a null-check and then call Double.parseDouble(String.valueOf(<RAW_VALUE>)) or something similar, with error catching for whatever ParseException that method throws in case a non-numeric value is provided. The JS code in ScriptHelper that's passing centerCustomProps from JS to java isnt doing any handling of data type, and i think this could give you weird behaviors. I would suggest string casting and re-parsing like I suggest above if you use this route.

Second, I would not expect all of the values from the standard "setScriptOptions()" to automatically be passed to the java trigger layer. If you went with that route, you would need to additionally add a line in core EHR's blood.js onUpsert() code to call something to the effect of: 
```
getJavaHelper().doWarnForBloodNearOverages(triggerHelper.doWarnForBloodNearOverages())
getJavaHelper().setBloodNearOverageThreshold(triggerHelper.getBloodNearOverageThreshold())
```
In other words, the scriptProperties are only copied from the JS into the java layer if some code actually does it. Most of them are never used in the java code, or are only copied in when actually needed. The primary reason the scriptProperties object exists is to get at least some documentation and minimal discoverability around what config options are supported without needing to read into the depths of the code. 

Anyway, what you're doing with setCenterCustomProps is at least as valid as some of the other things going through it.

Reply:
where do you see it being passed as Map<String, String>? I changed it from Map<String,String> to Map<String, Object> _centerCustomProps; and those doubles are still getting sent over as Integers, so I did something like:
```
Object theVal = _centerCustomProps.get("bloodNearOverageThreshold");
if (theVal instanceof Integer)
{
  Integer theValInt = (Integer) theVal;
  return theValInt.doubleValue();
}```

Reply:
OK, apologies. I must have been mixing up what I looked at locally (which doesnt have your changes) with this. 

That all being said, my experience with dealing with JS objects being ported into java is that it's really easy to get obscure bugs, I so I generally avoid it. For example, you might have issues with Rhino casting the JS integer/double into Number or some different java classes. I am not running your code locally, so I am just saying this from experience, not testing your code. Anyway, if you think you need to do 'instanceof' tests, I strongly recommend doing what I suggested above and using Double.parseDouble(String.valueOf(<theValue>). It's not really any less awkward than instanceOf testing, and should be robust to however rhino decides to wrap the JS data.

Reply:
Also, while your example is 4.0, and thus double->int->double doesnt matter, would your Integer casting actually work if the value wasnt an integer?

Reply:
no worries I appreciate the feedback. So As I understand String.valueOf() expects a string, but if my Map<String, Object> values are "objects" it looks like Rhino or whatever is doing the translation sometimes works, in that if I pass in a boolean, it gets casted in as a boolean for the most part, but passing in a double of 4.0 it gets casted to an Integer then I have to cast it back to an int, otherwise just use the default value.

so the full codeblock would be something like:

```
 public double getBloodNearingOveragesThreshold()
    {
        Object theVal = _centerCustomProps.get("bloodNearOverageThreshold");
        if (null != theVal)
        {
            if (theVal instanceof Integer)
            {
                Integer theValInt = (Integer) theVal;
                return theValInt.doubleValue();
            }
            else if (theVal instanceof Double)
            {
                return (Double) theVal;
            }
        }
        return _bloodNearingOveragesThresholdDefaultValue;
    }
    ```

I am not sure how much data checking should be done here, but maybe I can just add it to the documentation to say this prop's value is expected to be a double?

Reply:
To break down those points:
- My impressions of rhino could have bene fixed in newer versions, but in the past I found it to be inconsistent as far as how it translates JS objects into java, and quirky variations in your JS objects would change the result on the java side. Therefore I generally trust it as little as possible, and untyped Map<String, Object>s or similar should be avoided here, in my opinion.

- You wrote: "so As I understand String.valueOf() expects a string" - no, it does not and that's exactly why I propose it here. The reason I'm suggesting that pattern is that I think it's the most robust serialization that will be tolerant to whatever quirky variations happen in your data. 

Anyway, what you have is fine - it's not worth going back and forth. I would suggest adding an error check:
```
            else if (theVal instanceof Double)
            {
                return (Double) theVal;
            }
else {
    // Either throw or log an error here
}

```

Reply:
yeah, sorry i meant the variant of Double.valueOf() expects strings. The reason I chose the untyped Map<String,Object> was because it seemed odd to go from double -> string -> double. 

But anyway I have an even bigger problem with this setScriptOptions codepath (in this case setCenterCustomProps). Since EHR's blood.js does not have a registered INIT trigger, we cannot override it with one. Therefore we need to override it only with an onInit which means duplicating all of the code in the current onInit just to set some custom options for WNPRC. Can you please clarify this behavior @labkey-martyp and let us know what is the best approach for setting a custom "script option" for this usecase? Before going this route I just had exposed a public setter method in the trigger script helper that seemed the fastest way, but setting these script options is now causing me headaches.

Reply:
why do you need to override it? can't you just register you handler (which should be called after core EHR's onInit), and set whatever you need?

Reply:
my fault, I was registering the INIT handler incorrectly, but I was not getting any error messages so I was confused about what was going wrong. It looks to be working now with those custom script options. Throwing an error message now too if a null val is found.

Any other requested changes for this PR?